File: libpretixsync/src/test/java/eu/pretix/libpretixsync/db/QuestionAnswerRequiredValidationTest.java
package eu.pretix.libpretixsync.db;

import eu.pretix.libpretixsync.check.QuestionType;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;


public class QuestionAnswerRequiredValidationTest {

    @Test
    public void testBoolean() throws AbstractQuestion.ValidationException {
        Question q = new Question();
        q.setType(QuestionType.B);
        q.setRequired(true);
        q.clean_answer("True", q.getOptions());
        try {
            q.clean_answer("False", q.getOptions());
            fail("Expected an ValidationException to be thrown");
        } catch (Question.ValidationException e) {
        }
    }

    @Test
    public void testText() throws AbstractQuestion.ValidationException {
        Question q = new Question();
        q.setType(QuestionType.T);
        q.setRequired(true);
        q.clean_answer("True", q.getOptions());
        try {
            q.clean_answer("", q.getOptions());
            fail("Expected an ValidationException to be thrown");
        } catch (Question.ValidationException e) {
        }
    }
}

File: libpretixsync/src/test/java/eu/pretix/libpretixsync/db/QuestionAnswerRequiredValidationTest.kt
package eu.pretix.libpretixsync.db

import eu.pretix.libpretixsync.check.QuestionType
import org.junit.Assert.fail
import org.junit.Test
import eu.pretix.libpretixsync.models.Question as QuestionModel


class QuestionAnswerRequiredValidationTest {

    @Test
    @Throws(QuestionLike.ValidationException::class)
    fun testBoolean() {
        val q = QuestionModel(
            question = "Test",
            type = QuestionType.B,
            required = true,
            identifier = "ABTBAB8S",
            eventSlug = "test",
            id = 1L,
            serverId = 1L,
            position = 0L,
        )

        q.clean_answer("True", q.options, false)

        try {
            q.clean_answer("False", q.options, false)
            fail("Expected an ValidationException to be thrown")
        } catch (e: QuestionLike.ValidationException) {
        }

    }

    @Test
    fun testBooleanWithOptionalOverride() {
        val q = QuestionModel(
            question = "Test",
            type = QuestionType.B,
            required = true,
            identifier = "ABTBAB8S",
            eventSlug = "test",
            id = 1L,
            serverId = 1L,
            position = 0L,
        )

        try {
            q.clean_answer("False", q.options, true)
        } catch (e: QuestionLike.ValidationException) {
            fail("Expected that a ValidationException is not thrown")
        }
    }

    @Test
    @Throws(QuestionLike.ValidationException::class)
    fun testText() {
        val q = QuestionModel(
            question = "Test",
            type = QuestionType.T,
            required = true,
            identifier = "ABTBAB8S",
            eventSlug = "test",
            id = 1L,
            serverId = 1L,
            position = 0L,
        )

        q.clean_answer("True", q.options, false)
        q.clean_answer("False", q.options, false)
        try {
            q.clean_answer("", q.options, false)
            fail("Expected an ValidationException to be thrown")
        } catch (e: QuestionLike.ValidationException) {
        }
    }

    @Test
    fun testTextWithOptionalOverride() {
        val q = QuestionModel(
            question = "Test",
            type = QuestionType.T,
            required = true,
            identifier = "ABTBAB8S",
            eventSlug = "test",
            id = 1L,
            serverId = 1L,
            position = 0L,
        )

        try {
            q.clean_answer("", q.options, true)
        } catch (e: QuestionLike.ValidationException) {
            fail("Expected that a ValidationException is not thrown")
        }
    }
}